solver/src/interactions/uniswap.rs

```diff
@@ -1,33 +1,54 @@
-use crate::{encoding, settlement::Interaction};
+use super::encode_interaction;
+use crate::settlement::Interaction;
 use anyhow::Result;
-use contracts::UniswapV2Router02;
+use contracts::{GPv2Settlement, UniswapV2Router02, IERC20};
 use primitive_types::{H160, U256};
 
 #[derive(Debug)]
 pub struct UniswapInteraction {
     pub contract: UniswapV2Router02,
+    pub settlement: GPv2Settlement,
+    pub set_allowance: bool,
     pub amount_in: U256,
     pub amount_out_min: U256,
     pub token_in: H160,
     pub token_out: H160,
-    pub payout_to: H160,
 }
 
 impl Interaction for UniswapInteraction {
     fn encode(&self, writer: &mut dyn std::io::Write) -> Result<()> {
+        self.encode_approve(writer)?;
+        self.encode_swap(writer)
```

Comment:
On the contract level the unsiwap router interaction and the approve interaction are different interaction. Here they are kind of in a common interaction struct.

Your choice makes sense, as these two are tightly coupled. But I would not have expected them under a common UniswaapInteraction just because they are different ones on the smart contract level.

Reply:
I do think they are very related. To me an interaction in the context of the solver is a recipe of how to use some on-chain liquidity (settle against a smart contract order). Each component should be responsible for deciding what it needs to realize the desired trade. In this case if we have never traded against that token on Uniswap it means encoding two interactions, otherwise one.

Similarly I'd see some other implementation of that trade dealing with Curve, Balancer or whatsoever. We can share the logic of how to encode an approval somewhere more general.

We could rename the method to `encode_interactions` as in the smart contract an interaction is one call and here one order can result in multiple interactions.